### ISSUE TITLE
Larve Mermaid SVG cut off

### ISSUE THREAD

**Issue by jeppevinkel** [yuzutech/kroki#1345]
Large Mermaid diagrams have parts that get cut off when generated as SVG.

This example diagram from Mermaid causes issues
```
classDiagram
Class01 <|-- AveryLongClass : Cool
Class03 *-- Class04
Class05 o-- Class06
Class07 .. Class08
Class09 --> C2 : Where am i?
Class09 --* C3
Class09 --|> Class07
Class07 : equals()
Class07 : Object[] elementData
Class01 : size()
Class01 : int chimp
Class01 : int gorilla
Class08 <--> C2: Cool label
```
```mermaid
classDiagram
Class01 <|-- AveryLongClass : Cool
Class03 *-- Class04
Class05 o-- Class06
Class07 .. Class08
Class09 --> C2 : Where am i?
Class09 --* C3
Class09 --|> Class07
Class07 : equals()
Class07 : Object[] elementData
Class01 : size()
Class01 : int chimp
Class01 : int gorilla
Class08 <--> C2: Cool label
```

**Comment by Mogztter** [comment]
I believe this is a font issue, here's how it looks like with Kroki:

![image](https://user-images.githubusercontent.com/333276/196000612-cae4c188-4595-4d1f-a37c-df43cd0ee0b8.png)

**Comment by Mogztter** [comment]
By default, Mermaid is using the following font: `"trebuchet ms",verdana,arial,sans-serif`.

**Comment by Mogztter** [comment]
Probably related to https://github.com/mermaid-js/mermaid/issues/1540
It might be better to use "Roboto" by default. My guess is that `"trebuchet ms",verdana,arial,sans-serif` will render differently on Linux, macOS and Windows.

**Comment by Mogztter** [comment]
For reference, here's how the font looks on the container:

<img src="https://kroki.io/mermaid/png/eNpdz0ELgjAYBuB7v-I9lqBUVpqIEfMYdOwQHaZ86GK6mhYU_vg0l0i3fc_evexLJa-qWPBM82LCumG-QNjYNvZP0q-DKrOvIgBTSpqIC6tN9OeVsTXUYBtjHhzHkG9oC9uOwJZt4SknTeAFxG50aYG5o7GJTIE3dAag-4PLajob0TG5UlqfLyBJBZV1zGs-LBSgEm8a8h2Iskaai-L2Z5nSQsrfUx9h_91-fUiekPwALmZc9A==">

As you can see, the font rendering is different and text is not cut off.
Not sure if there's a solution since as far as I know all fonts render differently depending on the OS. One idea would be to use something like https://github.com/vercel/satori to convert the (HTML) text to SVG (i.e., `<path>`)

**Comment by Mogztter** [comment]
Mermaid could use something like https://github.com/danmarshall/google-font-to-svg-path which relies on https://maker.js.org/ to convert all text to SVG path that would be neat!

**Comment by Mogztter** [comment]
I did some tests and unfortunately each solution has drawbacks.
I also feel like this "issue" should be fix upstream. I had decent results using https://github.com/shrhdk/text-to-svg with Noto Sans.

**Comment by codingluke** [comment]
I encountered this too! Would be great to have it fixed :) on mermaid.live and here on github it works correct. Anyway great tool!

**Comment by yuriykuzin** [comment]
I also have this issue with pie chart diagrams, while it's ok on mermaid.live.

<img width="1126" alt="Screenshot 2024-02-27 at 13 22 23" src="https://github.com/yuzutech/kroki/assets/1729556/725e0506-8616-4934-b9df-272ffa535624">

```
pie title iPhone Usage Distribution
    "Social Media" : 35
    "Calls" : 15
    "Email" : 20
    "Gaming" : 10
    "Music & Podcasts" : 10
    "Photography" : 5
    "Navigation" : 5
```

**Comment by ggrossetie** [comment]
It's the same issue.
Rendering Mermaid server-side will produce these inconsistencies since the label width/height computation relies on fonts and fonts do not render exactly the same on two machines.

**Comment by boris-moduscreate** [comment]
Given that font is a configureable option in Mermaid.js would you consider making it an option which can be passed to Kroki?

**Comment by boris-moduscreate** [comment]
Yeah I absolutely agree, but the question is who decides what font to use? Do we wait for mermaid to fix it? Does Kroki set a default you believe is best? Should the Kroki docker image have the mermaid default font installed into it? Should the end user be able to provide an override font (assuming it is available)? We self-host Kroki so we don't really care which option is chosen, but we would like to fix it sooner rather than later if possible. Ultimately it's your project and we will defer to your preferred approach here.

**Comment by nithom** [comment]
A workaround for us is to tweak the margin settings for generated mermaid diagrams in <https://github.com/yuzutech/kroki/blob/main/mermaid/src/task.js>.

Within `this.mermaidConfig`, you can add something like

```
themeCSS: '.nodeLabel { margin: 3px 20px 3px; }'
```

This probably still might not work for some fonts, though.

**Comment by s-h-a-d-o-w** [comment]
As far as I've seen in tests today, even specifying a custom font in mermaid usually doesn't work because it renders before the custom font has loaded. (There's a race condition - depends on how quickly the font can be loaded or how slow the rendering is.)

But if that gets fixed, the appropriate way of handling this seems to be: 
1. Install the custom font on the kroki server.
2. Optional: If rendering the output as SVG, load the font wherever you actually use the SVG code.

And as far as workarounds go, with kroki obviously usually rendering on Linux, while the user could be on another OS, I would recommend checking out [modern font stacks](https://github.com/system-fonts/modern-font-stacks) and picking a combination where the text is similarly wide across the targeted platforms.

**Comment by nadiamoe** [comment]
Sharing here my workarounds to this issue for in case these help anyone out there:

Workaround 1: Install `trebuchet MS` on your system. That's the font used by kroki, so if your system (browser) has it, the SVG will render properly.

Workaround 2: Specify a different font you already have. For kroki/mermaid specifically, this can be done via the `font-size` ([src](https://docs.kroki.io/kroki/setup/diagram-options/#_mermaid)) URL parameter, something like:
```
http://localhost:8081/mermaid/svg/b66gibberish?theme=dark&font-family=Open+Sans
```
That font should be available both in the kroki container and in your computer. Open Sans is a good candidate for this in my experience.
